Populate dependencies node subtrees for packages and projects

drewnoakes opened the pull request:
Fixes: https://github.com/NuGet/Home/issues/9580

Populates descendant tree items for top-level package and projects displayed in the "Dependencies" node of projects based on the .NET Project System (which in turn is based on CPS).

Top-level dependencies are populated via project evaluation and design-time build data. Nodes lower in the tree are lazily constructed only if and when the user expands the tree enough for them to be visible, or if the user initiates search in Solution Explorer.

Data is obtained from the `project.assets.json` file, and updated whenever it changes. This data is captured in an immutable `AssetsFileDependenciesSnapshot` object and published via Dataflow in response to other project system data sources.

Tree items are constructed via the following mechanisms:

- Each type of node in the tree has its own 'item' (eg. `ProjectReferenceItem`, `PackageReferenceItem`).
- Children of top-level dependencies are reported via `PackageReferenceAttachedCollectionSourceProvider` and `ProjectReferenceAttachedCollectionSourceProvider`, which are called for all `IVsHierarchy` items in the tree.
- Further children are modelled declaratively via a set of 'relations' (eg. `ProjectToPackageRelation`, `ProjectToProjectRelation`, `PackageToPackageRelation`).

For more information about the APIs used to populate these subtrees, see https://github.com/dotnet/project-system/blob/master/docs/repo/dependencies-node-roadmap.md#transitive-dependencies

![image](https://user-images.githubusercontent.com/350947/82172277-12f6a500-990d-11ea-9728-a6874581c5f6.png)

drewnoakes committed:
Set ResXFileCodeGenerator on resx files

drewnoakes committed:
Bump CPS version

drewnoakes committed:
Use LangVersion 8

drewnoakes committed:
Add comment about configure.ps1 in bootstrap.proj

drewnoakes committed:
Populate dependencies node subtrees for packages and projects

Populate descendant tree items for top-level package and projects displayed in the "Dependencies" node of projects based on the .NET Project System (which in turn is based on CPS).

Top-level dependencies are populated via project evaluation and design-time build data. Nodes lower in the tree are lazily constructed only if and when the user expands the tree enough for them to be visible, or if the user initiates search in Solution Explorer.

Data is obtained from the `project.assets.json` file, and updated whenever it changes. This data is captured in an immutable `AssetsFileDependenciesSnapshot` object and published via Dataflow in response to other project system data sources.

Tree items are constructed via the following mechanisms:

- Each type of node in the tree has its own 'item' (eg. `ProjectReferenceItem`, `PackageReferenceItem`).
- Children of top-level dependencies are reported via `PackageReferenceAttachedCollectionSourceProvider` and `ProjectReferenceAttachedCollectionSourceProvider`, which are called for all `IVsHierarchy` items in the tree.
- Further children are modelled declaratively via a set of 'relations' (eg. `ProjectToPackageRelation`, `ProjectToProjectRelation`, `PackageToPackageRelation`).

For more information about the APIs used to populate these subtrees, see https://github.com/dotnet/project-system/blob/master/docs/repo/dependencies-node-roadmap.md#transitive-dependencies

drewnoakes committed:
Add BrowseObjectBase implementation

This type was previously provided by the Project System as a workaround to the
fact that the shell's didn't have a method as virtual. That was changed in
https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/248337, and the type
made internal in https://github.com/dotnet/project-system/pull/6198.
Once the updated shell is available, `BrowseObjectBase` can be removed and
`LocalizableProperties` used directly in its place.

drewnoakes commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/NuGet.VisualStudio.Implementation.csproj:
```diff
@@ -9,10 +9,10 @@
     <Description>Implementation of the NuGet.VisualStudio extensibility APIs.</Description>
     <Guid>9623cf30-192c-4864-b419-29649169ae30</Guid>
     <ImportedFromTypeLib>NuGet.VisualStudio.Implementation</ImportedFromTypeLib>
+    <LangVersion>8</LangVersion>
```
This PR uses C#8 NRT. Is there a scenario where a compatible compiler would not be available?

nkolev92 replied:
I'd expect that it works everywhere. 
Locally we dogfood, build machines are latest 16.x stable.

donnie-msft replied:
I think this is ok.. @zivkan do you know ?

drewnoakes committed:
Bump Project System package versions

Requires updaing vs-threading and fixing a few warnings.

drewnoakes commented:
/azp run

nkolev92 commented:
We are not integrated with that :) 

I'll kick off a run.

donnie-msft commented:
@drewnoakes any Issue on your side or ours, or a design doc to explain the changes? We will need an Issue to link to this PR. Might need you or myself to create one...

nkolev92 commented:
@donnie-msft 

I tagged you in the issue, [1107557](https://dev.azure.com/devdiv/DevDiv/_workitems/edit/1107557)

donnie-msft commented:
Created and linked Issue. Will shepherd this in the coming days...

nkolev92 commented:
Thanks @donnie-msft I'll take a stab at a review soon.

drewnoakes commented:
@donnie-msft 

> a design doc to explain the changes

This document explains the design https://github.com/dotnet/project-system/blob/master/docs/repo/dependencies-node-roadmap.md#transitive-dependencies

I'm happy to answer specific questions or have an ad-hoc call to chat through the design while looking at the code.

drewnoakes committed:
Merge remote-tracking branch 'upstream/dev' into populate-dependencies-tree-items

drewnoakes commented:
Merged upstream to fix conflict.

nkolev92 commented on build/bootstrap.proj:
```diff
@@ -7,15 +7,20 @@
-        <PackageDownload Include="Microsoft.VisualStudio.ProjectSystem" version="[16.0.201-pre-g7d366164d0]" />
+        <PackageDownload Include="Microsoft.VisualStudio.Composition" version="[16.4.11]" />
+        <PackageDownload Include="Microsoft.VisualStudio.Imaging.Interop.14.0.DesignTime" version="[14.3.26930]" />
+        <PackageDownload Include="Microsoft.VisualStudio.ProjectSystem" version="[16.7.156-pre]" />
```
Can we add PackageReference instead? 
Not sure what the motivation was behind PackageDownloads for Composition and ProjectSystem, but I see no reason to not at least try to use a PackageReference.

drewnoakes replied:
For `Microsoft.VisualStudio.ProjectSystem` it's presumably to reference `Microsoft.VisualStudio.ProjectSystem.Interop` directly with `EmbedInteropTypes` set to `true`.

https://github.com/NuGet/NuGet.Client/blob/34ceb7d3d5a70ba4b1ecf196e5902dcc9ffed05d/test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/NuGet.PackageManagement.VisualStudio.Test.csproj#L62-L65

This isn't necessary for `Microsoft.VisualStudio.ProjectSystem.Managed` and `Microsoft.VisualStudio.ProjectSystem.Managed.VS`. I just tried converting to `PackageReference`, but that triggered a package downgrade error on `Microsoft.CodeAnalysis.FxCopAnalyzers`. Upgrading that to `3.0.0` triggers 90 new errors, many of which look like they're worth a review.

Filed https://github.com/NuGet/Home/issues/9620 to track this.

nkolev92 replied:
Yeah, that's fair :D

drewnoakes committed:
Ensure name/version are not empty

donnie-msft commented on build/bootstrap.proj:
```diff
@@ -7,15 +7,20 @@
+        <PackageDownload Include="Microsoft.VisualStudio.Imaging.Interop.14.0.DesignTime" version="[14.3.26930]" />
+        <PackageDownload Include="Microsoft.VisualStudio.ProjectSystem" version="[16.7.156-pre]" />
+        <PackageDownload Include="Microsoft.VisualStudio.ProjectSystem.Managed" version="[16.7.0-beta1-65318-02]" />
+        <PackageDownload Include="Microsoft.VisualStudio.ProjectSystem.Managed.VS" version="[16.7.0-beta1-65318-02]" />
```
Not sure here but is it a good idea for NuGet to depend on prerelease packages?

drewnoakes replied:
If it works like the Project System VSIX, then none of these references are actually used when packaging for VS. They're only used for compilation and local debugging.

Without using these prerelease packages, we won't be able to compile against the new as-yet-unreleased APIs.

donnie-msft commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/NullableAnnotations.cs:
```diff
@@ -0,0 +1,91 @@
+// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
```
> @zkat better to match the original dotnet license text, or ours?....
> ```suggestion
> // Copyright (c) .NET Foundation. All rights reserved.
> // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
> ```

We can't change copyright headers on other projects without their permission.

drewnoakes replied:
This file is originally MIT, copied from dotnet/runtime. I believe other repos leave this intact.

donnie-msft reviewed (commented):
more to review...

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Items/PackageReferenceItem.cs:
```diff
@@ -0,0 +1,88 @@
+#nullable enable
+
+using System.Diagnostics.CodeAnalysis;
+
```
We recently made a change to sort our usings without empty lines between classes of usings. Can you please reformat it to match that? The editorconfig should be enforcing it.

nkolev92 replied:
You might need to be rebase to get this change.

drewnoakes replied:
Update removes spaces. I merged `dev` but received no warnings so I don't think the editorconfig is working as you expect. I made the changes manually.

nkolev92 replied:
trying it locally, it only works the other direction, so it won't add spaces, but won't remove them.

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/AssetsFileDependenciesDataSource.cs:
```diff
@@ -0,0 +1,130 @@
+                    {
+                        if (path.Length != 0)
+                        {
+                            if (projectSnapshot.AdditionalDependentFileTimes == null || !projectSnapshot.AdditionalDependentFileTimes.TryGetValue(path, out timestampUtc))
```
NuGet has a need to have the assets file loaded in the PM UI. 
This is more of a self note to create an issue for this.

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/AssetsFileDependenciesTreeSearchProvider.cs:
```diff
@@ -0,0 +1,191 @@
+
+                continue;
+
+                async Task<ConfiguredProject?> FindConfiguredProjectAsync(string tfm)
```
From a style perspective we've avoided local functions, unless they are absolutely necessary. 

Before I suggest changing, I want to get @NuGet/nuget-client to chime in.

donnie-msft replied:
I already commented on this https://github.com/NuGet/NuGet.Client/pull/3392#discussion_r431171298

drewnoakes replied:
I don't think local function are ever _necessary_. I get that they're a relatively new language feature and folks aren't comfortable with them just yet.

What I like about local functions is they reduce a function's scope. Other languages have had this feature for a long time. You can see, for example, that function `CreateLogItem` is used only within the `SearchLogMessages` function. It means I tend to create a greater number of smaller functions, and their names become more self documenting. Flattening all these out to top level methods, I'd probably inline a bunch of them which I think hurts readability.

Another way of thinking about this is that we've generally agreed that variables should have as narrow scope as possible. Why should it be different for functions?

nkolev92 replied:
Good enough for me :)
Re-reading my comment, I'm not sure how I came up with `necessary` :)

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileDependenciesSnapshot.cs:
```diff
@@ -0,0 +1,214 @@
+            try
+            {
+                // Parse the file
+                using var fileStream = new FileStream(path, FileMode.Open, FileAccess.Read, FileShare.Read | FileShare.Delete, 4096 * 10, FileOptions.SequentialScan);
```
Can we use some of the NuGet APIs for reading with filestream? 

Like: https://github.com/NuGet/NuGet.Client/blob/5ae9cdac4033fb588c3235c961dc59055c4f16df/src/NuGet.Core/NuGet.ProjectModel/LockFile/LockFileUtilities.cs

drewnoakes replied:
Sure. Where would I get an `Common.ILogger` from?

nkolev92 replied:
Most of NuGet's operations require a more involved user gesture (like run PMC, run PM UI, run restore) and we have a logical answer for how to provide feedback to the user if something went wrong. 

Guess my questions, what if this throws? 
Is the error captured somehow? Shown to user/telemetry etc.

drewnoakes replied:
The current behaviour here is to no-op if an exception occurs. My assumption is that if the lock file fails to parse, the error will be surfaced in several other places, but I'm not 100% that's actually true.

We must catch the exception here as letting it escape will fault the dataflow block which stops all future updates.

Happy to add telemetry or whatever makes sense here in the NuGet code, just let me know.

nkolev92 replied:
I am fine with leaving as is right now.

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileDependenciesSnapshot.cs:
```diff
@@ -0,0 +1,214 @@
+
+            _packageFolders = lockFile.PackageFolders.Select(pf => pf.Path).ToImmutableArray();
+
+            ImmutableDictionary<string, AssetsFileTarget>.Builder dataByTarget = ImmutableDictionary.CreateBuilder<string, AssetsFileTarget>(StringComparer.OrdinalIgnoreCase); // TODO review comparer here -- should it be ignore case?
```
I think you can remove the TODO. It should be ignore case. 

NuGet has a platforawarecomparer method somewhere, maybe search for that for code consistency.

drewnoakes replied:
Thanks, removed the TODO.

I spent a few minutes searching for various names of a platform-aware comparer but couldn't find any definition or usage.

nkolev92 replied:
We can just keep ignore case. 
Looking through the rest of our code, we're not using that consistently and it's more relevant when dealing with paths explicitly. 

I also sent you on a wild goose chase with the platform aware :D
It's [PathUtility.GetStringComparerBasedOnOS](https://github.com/NuGet/NuGet.Client/blob/34ceb7d3d5a70ba4b1ecf196e5902dcc9ffed05d/src/NuGet.Core/NuGet.Common/PathUtil/PathUtility.cs#L21)

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileTarget.cs:
```diff
@@ -0,0 +1,192 @@
+        /// <summary>
+        /// Gets the target framework moniker, such as <c>.NETFramework,Version=v4.8</c> or <c>.NETStandard,Version=v1.3</c>.
+        /// </summary>
+        public string TargetFrameworkMoniker { get; }
```
Given that we are making comparisons based on this it makes a bit uneasy but I don't have better suggestions :(

drewnoakes replied:
This appears to be the identifier for targets in the lockfile itself. Is there something better?

nkolev92 replied:
I think eventually we'd need to use the inner build targetframeworks value but that's not where we are right now. 

I think we can resolve this conversation and mention this discussion in https://github.com/NuGet/Home/issues/5154.

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/AssetsFileDependenciesDataSource.cs:
```diff
@@ -0,0 +1,130 @@
+                PropagateCompletion = true  // Make sure source block completion and faults flow onto the target block to avoid hangs.
+            };
+
+            IProjectVersionedValue<AssetsFileDependenciesSnapshot> Transform(IProjectVersionedValue<ValueTuple<IProjectSnapshot, IProjectSubscriptionUpdate>> update)
```
This isn't called on a UI thread correct? I didn't get to debug yet, no P2 on the machine I am working on.

nkolev92 replied:
On a related note, is the project-system implementation completely removed now?

drewnoakes replied:
Yes the Project System side was completely removed for 16.7p1. This PR wasn't merged by the deadline (I only submitted it a handful of days beforehand), so functionality regressed in the public preview. I'm keen to restore the feature in 16.7p2 so we have more time to test everything before GA.

Do you have VS master on your machine? That's what I've been using to test all this stuff.

drewnoakes replied:
And no, this isn't called on the UI thread.

nkolev92 replied:
Yeah I just need to debug it on the machine that has master. 
The above 2 clarify my concerns though so resolving this thread.

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/AssetsFileTopLevelDependenciesCollectionSourceProvider.cs:
```diff
@@ -0,0 +1,144 @@
+
+        protected abstract bool TryGetIdentity(Properties properties, out TIdentity identity);
+
+        protected abstract bool TryGetLibrary(AssetsFileTarget target, TIdentity identity, [NotNullWhen(returnValue: true)] out AssetsFileTargetLibrary? library);
```
You mention that these will only get created when the node is expanded (not top level), so can you help me understand how it'll work in this case?
When is the target created?

drewnoakes replied:
The target is part of the immutable snapshot of data from the assets file. The snapshot is created lazily once per-project as soon as the first child needs it. The data source that produces these starts doing so when it is first subscribed to.

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Items/PackageContentFileItem.cs:
```diff
@@ -0,0 +1,90 @@
+        {
+            // Content file paths always start with "contentFiles/" so remove it from display strings
+            const string Prefix = "contentFiles/";
+            return rawPath.StartsWith(Prefix, StringComparison.Ordinal) ? rawPath.Substring(Prefix.Length) : rawPath;
```
It's probably safe to assume it'll contain it & avoid the comparsion.

drewnoakes replied:
I'd still want to have a length check to avoid `ArgumentOutOfRangeException`. Checking for the prefix isn't much more overhead.

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Items/ProjectReferenceItem.cs:
```diff
@@ -0,0 +1,77 @@
+    /// <summary>
+    /// Backing object for transitive project reference nodes in the dependencies tree.
+    /// </summary>
+    internal sealed class ProjectReferenceItem : RelatableItemBase
```
Aren't ProjectReferenceItem and PackageReference item identical in pretty much everything but display (browsable object). 
Can the classes be merged and that way we can simplify the relations? (or reduce the number of them)
Is it an overkill potentially?

drewnoakes replied:
There are different relations for each type. For example, we don't show "Framework Assemblies" for projects.

They also have different priorities, icons, context menus.

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Items/PackageReferenceItem.cs:
```diff
@@ -0,0 +1,88 @@
+
+        public override ImageMoniker IconMoniker => ManagedImageMonikers.NuGetGrey;
+
+        protected override bool TryGetProjectNode(IProjectTree targetRootNode, IRelatableItem item, [NotNullWhen(returnValue: true)] out IProjectTree? projectTree)
```
I really wish the APIs had docs :( It'd be easier to follow what's going on.

drewnoakes replied:
The base has docs. VS shows them on overrides. If I duplicate them here (and onto the many other item subclasses), they'll just fall out of date. I encourage you to review this in VS rather than on GitHub.

https://github.com/dotnet/project-system/blob/1587349ca69b35ce320f7af8f264e89ce12bdea6/src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Tree/Dependencies/AttachedCollections/IRelatableItem.cs#L71-L86

nkolev92 replied:
I was reviewing in VS and it showed nothing for me, weird. 
That'd helpful though, thanks.

drewnoakes replied:
Yeah I realised after writing that comment that this API was missing a hint, hence this PR: https://github.com/dotnet/project-system/pull/6251

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileTargetLibrary.cs:
```diff
@@ -0,0 +1,93 @@
+            Type = type;
+
+            // TODO use each dependency's version range in caption (won't have parity with top-level item unless we update caption or change SDK to return this information)
+            // TODO use each dependency's include/exclude in browse object (won't have parity with top-level item until we rethink browse objects for them)
```
Can we create follow up issues for these TODOs and link them here. 
I might have some follow up questions, cause I'm not sure I understand the concern.

drewnoakes replied:
https://github.com/NuGet/Home/issues/9616

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileTargetLibrary.cs:
```diff
@@ -0,0 +1,93 @@
+
+            CompileTimeAssemblies = library.CompileTimeAssemblies
+                .Select(a => a.Path)
+                .Where(path => path != null)
```
path shouldn't be null usually.

drewnoakes replied:
I don't remember why I added this now. I'm reluctant to take it out without doing a whole bunch of testing, so will leave it.

nkolev92 replied:
That's fair. I think the compiler will try to optimize the consecutive where statements :)

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileTargetLibrary.cs:
```diff
@@ -0,0 +1,93 @@
+
+            FrameworkAssemblies = library.FrameworkAssemblies.ToImmutableArray();
+
+            // TODO filter by code language as well (requires knowing project language): https://github.com/dotnet/NuGet.BuildTasks/blob/5244c490a425353ac12445567d87d674ae118836/src/Microsoft.NuGet.Build.Tasks/ResolveNuGetPackageAssets.cs#L572-L575
```
Same here, we should create an issue.

drewnoakes replied:
https://github.com/NuGet/Home/issues/9615

drewnoakes replied:
https://github.com/NuGet/Home/issues/9617

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileTargetLibrary.cs:
```diff
@@ -0,0 +1,93 @@
+        public AssetsFileLibraryType Type { get; }
+        public ImmutableArray<string> Dependencies { get; }
+        public ImmutableArray<string> FrameworkAssemblies { get; }
+        public ImmutableArray<string> CompileTimeAssemblies { get; }
```
We are missing the Framework References here, should we add them? 
Specifically I am referring to shared frameworks in the .NET Core world.

drewnoakes replied:
https://github.com/NuGet/Home/issues/9618

nkolev92 commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/AssetsFileTopLevelDependenciesCollectionSourceProvider.cs:
```diff
@@ -0,0 +1,144 @@
+
+            IProjectThreadingService projectThreadingService = unconfiguredProject.Services.ThreadingPolicy;
+
+            projectThreadingService.VerifyOnUIThread();
```
What's the difference between this and ThreadHelper?

drewnoakes replied:
We have banned `ThreadHelper` in the .NET Project System because it's static and prohibits unit testing.

https://github.com/dotnet/project-system/blob/1587349ca69b35ce320f7af8f264e89ce12bdea6/src/Common/BannedSymbols.txt#L82

drewnoakes committed:
Produce context menus for some transitive items

https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/250942 adds
these menus for use on transitive assemblies, packages and projects.
This allows various teams to attach commands to these tree items.

donnie-msft commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/AssetsFileDependenciesTreeSearchProvider.cs:
```diff
@@ -0,0 +1,191 @@
+
+                SearchLogMessages();
+
+                continue;
```
I suppose this is to optimize to avoid touching the following delegates. But, I'm not sure I'm a fan of so many delegates defined inside a loop. 
I don't see anything in our coding guidelines against it. I'm ok to leave it if @zivkan & @nkolev92 don't have other thoughts.

drewnoakes replied:
This isn't an optimisation, it's a style thing. Roslyn does it too. The idea is that local functions are towards the end and you put a `return` (in this case, continue) to demarcate the boundary.

The local functions below are not delegates. In fact the compiler can generate better code than it would for delegates, using a value type for enclosed state where possible.

donnie-msft commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/BrowseObjectBase.cs:
```diff
@@ -0,0 +1,70 @@
+    /// right from it)
+    /// </para>
+    /// <param>
+    /// PR https://dev.azure.com/devdiv/DevDiv/_git/VS/pullrequest/248337 makes the method
```
Can you file an issue in our Repo explaining what we need to do that is linked here via a TODO so we can track this?

drewnoakes replied:
It's on my list to fix this in three repos once the updated package is available.

The change is to remove this file completely and replace instances of `BrowseObjectBase` with `LocalizableProperties`.

drewnoakes replied:
https://github.com/NuGet/Home/issues/9619

donnie-msft commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/BrowseObjectDescriptionAttribute.cs:
```diff
@@ -0,0 +1,40 @@
+    /// Specifies a localized description for a property or event.
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Property)]
+    internal sealed class BrowseObjectDescriptionAttribute : DescriptionAttribute
```
For something that's duplicated like this, how do we ensure our code stays current with VS?
Is it as simple as a comment, or something more complex?

Appears copied from [Microsoft.VisualStudio.ProjectSystem.VS.Tree.Dependencies.AttachedCollections.Implementation](http://ddindex/?leftProject=Microsoft.VisualStudio.ProjectSystem.Managed.VS&file=ProjectSystem%5CVS%5CTree%5CDependencies%5CAttachedCollections%5CImplementation%5CBrowseObjectDescriptionAttribute.cs)

drewnoakes replied:
You don't have to keep it in sync.

Also it's not an exact copy. This attribute binds to your own resource dictionary, `VsResources`.

donnie-msft commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Items/AttachedItemPriority.cs:
```diff
@@ -0,0 +1,27 @@
+namespace NuGet.VisualStudio.SolutionExplorer
+{
+    /// <summary>
+    /// Specifies the order of attached items in the dependencies tree.
```
If this ordering changes in the Platform... does NuGet now override that ordering? 
This is another sort of thing that worries me of duplicating in our VSIX.

drewnoakes replied:
This ordering is local to this code. Only NuGet will produce these items. If another provider also attaches items to these parents (none currently do) then they'll have to fit in with these priority values. It's no different to priorities in a `.vsct` file, for example.

These values are not defined in the .NET Project System. What duplication are you referring to?

donnie-msft commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Items/MenuController.cs:
```diff
@@ -0,0 +1,62 @@
+
+namespace NuGet.VisualStudio.SolutionExplorer
+{
+    internal sealed class MenuController : IContextMenuController
```
Mostly asking for future work on my end.... We have other menu GUIDs defined in our VSCT (eg [IDM_VS_CTXT_PACKAGEREFERENCE_GROUP](https://github.com/NuGet/NuGet.Client/blob/dev/src/NuGet.Clients/NuGet.Tools/NuGetTools.vsct#L270)). Should we move those to this interface & pattern?

drewnoakes replied:
I don't think so. This code only selects which menu should be chosen for these items. Your `.vsct` file may in future work with these menus to add commands to transitive items (eg. converting a transitive package reference to a direct package reference), but that work would be separate.

donnie-msft commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileDependenciesSnapshot.cs:
```diff
@@ -0,0 +1,214 @@
+
+            _packageFolders = lockFile.PackageFolders.Select(pf => pf.Path).ToImmutableArray();
+
+            ImmutableDictionary<string, AssetsFileTarget>.Builder dataByTarget = ImmutableDictionary.CreateBuilder<string, AssetsFileTarget>(StringComparer.OrdinalIgnoreCase); // TODO review comparer here -- should it be ignore case?
```
Who owns this TODO? We only allow TODO's with linked issues, so please create one against us and either provide details, or reference the blocking issue on another repo.

drewnoakes replied:
Removed.

donnie-msft commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileDependenciesSnapshot.cs:
```diff
@@ -0,0 +1,214 @@
+            }
+        }
+
+        public override string ToString() => $"{DataByTarget.Count} target{(DataByTarget.Count == 1 ? "" : "s")}";
```
```suggestion
        public override string ToString() => $"{DataByTarget.Count} target{(DataByTarget.Count == 1 ? string.Empty : "s")}";
```

donnie-msft commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileTarget.cs:
```diff
@@ -0,0 +1,192 @@
+        {
+            var s = new StringBuilder();
+            s.Append("Target \"").Append(TargetFrameworkMoniker).Append("\" ");
+            s.Append(LibraryByName.Count).Append(LibraryByName.Count == 1 ? " library" : " libraries");
```
Localization?

drewnoakes replied:
This is only for debugging. No customer would ever see this string.

donnie-msft commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileTargetLibrary.cs:
```diff
@@ -0,0 +1,93 @@
+            Version = library.Version.ToNormalizedString();
+            Type = type;
+
+            // TODO use each dependency's version range in caption (won't have parity with top-level item unless we update caption or change SDK to return this information)
```
TODO's need issues (see earlier comment)

donnie-msft reviewed (commented):
Seems like a whole lot of duplicated classes are being introduced. Left a few comments, but can you address the big picture of this duplication and the strategy keeping repos in sync?

Also, should we any Tests be added on our end for any of this? The NuGet specific stuff seems should have a few tests at least. Maybe the bulk of this being VS code doesn't need testing on our side, but perhaps a few sanity checks to identify when VS changes regressed us?

nkolev92 reviewed (commented):
Looks great overall. 

I have a bunch of questions to further my understand of the implementation.
Also looks like it's pretty minimalistic from a data model perspective, everything relevant is contained within the snapshots, and the assets file is not kept in memory (correct me if I'm wrong).

Given that these snapshots stay in memory pretty much the whole time, have you looked at the size at all?

drewnoakes commented on src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileTargetLibrary.cs:
```diff
@@ -0,0 +1,93 @@
+                .Select(a => a.Path)
+                .Where(path => path != null)
+                .Where(path => !IsPlaceholderFile(path))
+                .ToImmutableArray(); // TODO do we want to use the 'properties' here? maybe for browse object
```
https://github.com/NuGet/Home/issues/9617

drewnoakes committed:
Remove TODO comments

These have been moved to issues on NuGet.

drewnoakes committed:
Merge remote-tracking branch 'upstream/dev' into populate-dependencies-tree-items

drewnoakes committed:
Remove space between using directive groups

drewnoakes committed:
Use string.Empty

drewnoakes commented:
@donnie-msft 

> Seems like a whole lot of duplicated classes are being introduced.

I assume you're referring to base types which also exist on the .NET Project System side. I originally exposed these, but @davkean rightfully pointed out that these types are nothing to do with the project system, but rather the framework's component model. They shouldn't be public API in the project system.

> can you address the big picture of this duplication and the strategy keeping repos in sync?

There's no need to keep these in sync.

> should we any Tests be added on our end for any of this? The NuGet specific stuff seems should have a few tests at least. Maybe the bulk of this being VS code doesn't need testing on our side, but perhaps a few sanity checks to identify when VS changes regressed us?

What sort of tests would you expect? Integration tests? We have some integration tests that could cover this on the .NET Project System side.

---

@nkolev92 

> looks like it's pretty minimalistic from a data model perspective, everything relevant is contained within the snapshots, and the assets file is not kept in memory (correct me if I'm wrong).

Yep, the `LockFile` object graph is collected once the snapshot is constructed.

> Given that these snapshots stay in memory pretty much the whole time, have you looked at the size at all?

Yes the snapshot is retained, as it must be available when the user expands tree items in order to populate them. It's only loaded once the dependencies tree is expanded enough to start showing dependencies (or if the user searches in Solution Explorer).

I've compared total memory usage across VS before/after this (and the .NET Project System) change. For `Roslyn.sln` we save about 200MB on the managed heap after solution load.

Further improvements are possible, for example reusing strings from the last snapshot when building the new one (so older strings aren't freed for higher gen collections), but it's not high enough priority for me to go after right now.

nkolev92 reviewed (approved):
LGTM. 
Thanks @drewnoakes 

Given the lack of too testing I think we should create follow up issues to implement tests. 

* We should try to add as many unit tests as possible. Many of the types can be tested in isolation (for example, DisposableBag etc). 
* Apex tests if possible

However in the mean time, we should try to get some manual validation on this. 
@donnie-msft as the shepherd can you help with that? 

@drewnoakes 
Can you please create tracking issues for the above 2.

drewnoakes commented:
> Can you please create tracking issues for the above 2.

Done.
